Prevent silent auth cycle on authentication error

NSeydoux opened the pull request:
When an authentication error happens, local storage is now cleared in order to prevent further silent login attempts. This should fix the issue where failed login cycle endlessly.

In order to implement this, the Session object listens on authentication events bubbled up by the handlers. In particular, on login, the current session ID is stored in local storage to enable silent auth in the future, and it is cleared on logout/auth error.

- [X] I've added a unit test to test for potential regressions of this bug.
- [x] The changelog has been updated, if applicable.
- [X] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).

NSeydoux committed:
Prevent silent auth cycle on authentication error

When an authentication error happens, local storage is now cleared in order to prevent further silent login attempts. This should fix the issue where failed login cycle endlessly.

In order to implement this, the Session object listens on authentication events bubbled up by the handlers. In particular, on login, the current session ID is stored in local storage to enable silent auth in the future, and it is cleared on logout/auth error.

ThisIsMissEm commented on packages/browser/src/Session.spec.ts:
```diff
@@ -922,6 +922,44 @@ describe("Session", () => {
+      // The local storage should have been cleared by the auth error
+      expect(window.localStorage.getItem(KEY_CURRENT_SESSION)).toBeNull();
+      // Silent authentication should not have been attempted
+      expect(clientAuthentication.login).not.toHaveBeenCalled();
```
I think we also want to assert here that `mySession.info.isLoggedIn` is false, too

NSeydoux replied:
Fixed in 3be5eb8b80dd725d59a031b4f9fa76ea5f76c048

ThisIsMissEm commented on packages/browser/src/Session.ts:
```diff
@@ -227,6 +227,27 @@ export class Session extends EventEmitter {
+      // Store the current session ID specifically in 'localStorage' (i.e., not using
+      // any other storage mechanism), as we don't deem this information to be
+      // sensitive, and we want to ensure it survives a browser tab refresh.
+      window.localStorage.setItem(KEY_CURRENT_SESSION, this.info.sessionId);
```
Is there a reason to not use the SessionInfoManager here? Or at least, to wrap the interaction with localstorage in a class?

NSeydoux replied:
The reason for it being separate from the `SessionInfoManager` is because this is a very browser-specific mechanism, which doesn't make sense in a browser context: it is only useful for silent login on refresh. `SessionInfoManager`, on the other hand, is the abstraction layer on top of storage which is shared between the browser and node codebase.

It could be wrapped in a separate class, but the intent is that it is basically scoped to the `Session` class, so I'm not sure the extra indirection is required.

ThisIsMissEm replied:
Right, it's just that we're not able to mock it as a clear interface (we're mocking localstorage instead) and we're duplicating keys, plus there's a few other localStorage interactions that happen directly, and they're not all in one file.

NSeydoux replied:
Yeah, some of these localStorage interaction are only required for a deprected workaround, and I'm currently investigating to see if anyone is still relying on it (I don't think so, and if not I'd want to remove it altogether). Otherwise, I think the only significant localStorage interaction is in `ClientAuthentication`, and I'll move it to the `Session` file so that it is all located in one place.

I don't know if there would be clear benefits in mocking it as a different interface, as this is explicitly relying on the browser local storage, as opposed to the `StorageUtility` we use otherwise, which is injected a storage instead.

ThisIsMissEm commented on packages/browser/src/Session.ts:
```diff
@@ -227,6 +227,27 @@ export class Session extends EventEmitter {
+    });
+
+    // On logout, silent refresh should no longer be considered a viable option.
+    this.on("logout", () => {
```
Question: do we reliably have a "session expired" event? If so, we should also remove there which might assist in preventing the `client_invalid` errors

NSeydoux replied:
That's a good point, and we do. It may save a round trip in some cases, even though the event being thrown is quite unlikely: either the browser tab where the refresh happens is in focus, and refresh happens automatically in the background, or it is out of focus, the refresh doesn't happen in the background, but sending the session expired signal doesn't happen either (or at least this is my current understanding). 

Regardless, session expiration is a possible event, and it should be handled here.

ThisIsMissEm replied:
hmm, is there maybe a way of tracking when a refresh should've happened, and if it didn't on refocus of the tab (https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API) we can detect that the session has potentially expired and we trigger a "refresh the session now", and then clear if it's returning `interaction_required`?

NSeydoux replied:
Fixed in cadb21e7095c3d020f29d6146c7cd7e5ae1561fa and 124c004311e95d3de470bc408732cfc469cb1d80

ThisIsMissEm commented on packages/browser/src/Session.ts:
```diff
@@ -227,6 +227,27 @@ export class Session extends EventEmitter {
+    // to prevent endless redirection, and session information should be cleared.
+    this.on(EVENTS.ERROR, async () => {
+      window.localStorage.removeItem(KEY_CURRENT_SESSION);
+      await this.logout();
```
I wasn't sure in here if calling `logout()` was correct — I think we actually want a "logout" that doesn't come from user interaction or SDK API call, that way we can differentiate a logout from a session had an error?

NSeydoux replied:
I think a consuming app may already distinguish between the two, by registering different callbacks to `onError` vs `onLogout`. However, it is true that calling `logout` here will also send the logout signal. I'm not sure if it is a wrong thing or not...

ThisIsMissEm replied:
I feel like it may be?

NSeydoux replied:
a66d20282c4d4abab963a65a6449e6e364ede7ba should clean this up. The logout signal is now only sent in case of user-initiated logout, and all the logout cleanup is located in one place.

ThisIsMissEm commented on packages/browser/src/login/oidc/redirectHandler/AuthCodeRedirectHandler.spec.ts:
```diff
@@ -456,11 +455,6 @@ describe("AuthCodeRedirectHandler", () => {
       );
-      // Check that the current session is stored correctly __specifically__ in
-      // 'localStorage'.
-      expect(window.localStorage.getItem(KEY_CURRENT_SESSION)).toBe(
```
Why isn't this check needed? Should the assertion be covered in a different unit test?

NSeydoux replied:
That's right, it has been moved from `AuthCodeRedirectHandler` to `Session`. I'll cover this in an appropriate unit test.

NSeydoux replied:
Fixed in 124c004311e95d3de470bc408732cfc469cb1d80

ThisIsMissEm commented on packages/browser/src/Session.spec.ts:
```diff
@@ -922,6 +922,44 @@ describe("Session", () => {
+      // eslint-disable-next-line no-void
+      void mySession.handleIncomingRedirect("https://some.redirect/url");
+      // The local storage should have been cleared by the auth error
+      expect(window.localStorage.getItem(KEY_CURRENT_SESSION)).toBeNull();
```
I can't see a test that actually checks this value transitioning state, that might be a good unit test to have?

NSeydoux replied:
Fixed in 124c004311e95d3de470bc408732cfc469cb1d80

NSeydoux committed:
Verify that local storage is updated on login

NSeydoux committed:
Use constants for all events

NSeydoux committed:
Verify local storage cleared on session expiration

NSeydoux committed:
Updated changelog

ThisIsMissEm commented on packages/browser/src/Session.ts:
```diff
@@ -227,6 +227,31 @@ export class Session extends EventEmitter {
+    });
+
+    this.on(EVENTS.SESSION_EXPIRED, () => {
+      window.localStorage.removeItem(KEY_CURRENT_SESSION);
```
Does this also need to `await this.logout()` or is error emitted afterwards?

NSeydoux replied:
Fixed in a66d20282c4d4abab963a65a6449e6e364ede7ba, the error wasn't related to an actual logout indeed.

ThisIsMissEm commented on packages/browser/src/Session.spec.ts:
```diff
@@ -922,6 +940,70 @@ describe("Session", () => {
+      });
+
+      const mySession = new Session({ clientAuthentication });
+      mySession.emit(EVENTS.SESSION_EXPIRED);
```
Can we test this by actually inducing the expiry in mySession, instead of injecting the event? Just thinking if the code that produces this event may emit other events or change state in another way?

NSeydoux replied:
I think it would be possible to mock this by actually expiring the session, by mocking a token response with zero time to live, but that would be more involved. Also, the portion of the code emitting this signal is meant to be isolated: it's the closure where the tokens/keys are kept. The signal is meant to be the only way for this closure to change state. Other signals could be emitted, but they should not impact the session state.

ThisIsMissEm commented on packages/core/src/constant.ts:
```diff
@@ -32,10 +32,13 @@ export const SOLID_CLIENT_AUTHN_KEY_PREFIX = "solidClientAuthn:";
 
 export const EVENTS = {
-  NEW_REFRESH_TOKEN: "newRefreshToken",
   ERROR: "onError",
```
this ones a tad odd? Can we fix it to just be `"error"` ?

NSeydoux replied:
Yeah I was thinking that too, and I was initially reluctant because that's technically a public API, but it's not documented anywhere (people are expected to use the `onError` callback instead), so I'll go ahead and rename this.

ThisIsMissEm reviewed (commented):
This is looking good, I've several more questions though!

NSeydoux committed:
Rename event

NSeydoux committed:
unrename event

NSeydoux committed:
Move all localStorate ref in Session

To cleanup code organization, as much localStorage references as possible are moved at the top-level in the Session class.

NSeydoux committed:
Refactor logout

Session now has an internal logout method which can either send a logout signal or not. The signal is only meant to be sent on user-initiated logout, but the function is also called in the case of auth error or session expiration.

ThisIsMissEm commented:
The problem with onError is that there's no way to remove the listener, which is actually needed in the context of UI frameworks.

I think we'd be better exposing the on/off interface directly.

> On 3. Mar 2022, at 11:20, Zwifi ***@***.***> wrote:
> 
> ﻿
> @NSeydoux commented on this pull request.
> 
> In packages/core/src/constant.ts:
> 
> > @@ -32,10 +32,13 @@ export const SOLID_CLIENT_AUTHN_KEY_PREFIX = "solidClientAuthn:";
>  export const PREFERRED_SIGNING_ALG = ["ES256", "RS256"];
>  
>  export const EVENTS = {
> -  NEW_REFRESH_TOKEN: "newRefreshToken",
>    ERROR: "onError",
> Yeah I was thinking that too, and I was initially reluctant because that's technically a public API, but it's not documented anywhere (people are expected to use the onError callback instead), so I'll go ahead and rename this.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android. 
> You are receiving this because your review was requested.

-- 
This e-mail, and any attachments thereto, is intended only for use by the 
addressee(s) named herein and may contain legally privileged, confidential 
and/or proprietary information. If you are not the intended recipient of 
this e-mail (or the person responsible for delivering this document to the 
intended recipient), please do not disseminate, distribute, print or copy 
this e-mail, or any attachment thereto. If you have received this e-mail in 
error, please respond to the individual sending the message, and 
permanently delete the email.

ThisIsMissEm commented:
Wait, what? How so?

> On 3. Mar 2022, at 11:42, Zwifi ***@***.***> wrote:
> 
> ﻿
> @NSeydoux commented on this pull request.
> 
> In packages/core/src/constant.ts:
> 
> > @@ -32,10 +32,13 @@ export const SOLID_CLIENT_AUTHN_KEY_PREFIX = "solidClientAuthn:";
>  export const PREFERRED_SIGNING_ALG = ["ES256", "RS256"];
>  
>  export const EVENTS = {
> -  NEW_REFRESH_TOKEN: "newRefreshToken",
>    ERROR: "onError",
> Aha actually error is an invalid event name, which is probably why it was named as such in the first place. I totally forgot about this, I'll add a comment for the ever forgetful future me.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android. 
> You are receiving this because your review was requested.

-- 
This e-mail, and any attachments thereto, is intended only for use by the 
addressee(s) named herein and may contain legally privileged, confidential 
and/or proprietary information. If you are not the intended recipient of 
this e-mail (or the person responsible for delivering this document to the 
intended recipient), please do not disseminate, distribute, print or copy 
this e-mail, or any attachment thereto. If you have received this e-mail in 
error, please respond to the individual sending the message, and 
permanently delete the email.

NSeydoux commented:
If you run the tests from https://github.com/inrupt/solid-client-authn-js/pull/2002/commits/4a1566b2d12ab55732e7591a9069c9af1a0a8fb6, an exception is thrown complaining about an "invalid event name"

NSeydoux committed:
Fix linting

NSeydoux committed:
Rerename error event

NSeydoux commented on packages/core/src/constant.ts:
```diff
@@ -32,10 +32,14 @@ export const SOLID_CLIENT_AUTHN_KEY_PREFIX = "solidClientAuthn:";
 
 export const EVENTS = {
-  NEW_REFRESH_TOKEN: "newRefreshToken",
+  // `error` is not a valid event name, hence the `onError`.
```
I updated the comment to reference this docs page, because future me will need a reminder at some point.

NSeydoux committed:
Listen to error event in core

ThisIsMissEm commented on packages/core/src/constant.ts:
```diff
@@ -32,10 +32,14 @@ export const SOLID_CLIENT_AUTHN_KEY_PREFIX = "solidClientAuthn:";
 
 export const EVENTS = {
+  // Note that an `error` events MUST be listened to: https://nodejs.org/dist/latest-v16.x/docs/api/events.html#error-events.
+  ERROR: "error",
```
Where we emit error events, you can also use `emitter.listenerCount("error")` before actually emitting
